### PR TITLE
feat: add ChatGPT Pro/Plus subscription support (OpenAI Codex OAuth)

### LINKED ISSUE #269: Add ChatGPT Pro/Plus subscription support (OpenAI Codex OAuth)
## Summary

Add support for connecting a ChatGPT Pro/Plus subscription via OAuth, similar to the existing Anthropic Claude Pro/Max integration. This allows users to use their ChatGPT subscription to power Frontman without needing a separate OpenAI API key.

## Background

We already have Anthropic OAuth working (PKCE flow with manual code copy/paste). OpenAI has a similar OAuth flow for their "Codex" API endpoint that allows ChatGPT Pro/Plus subscribers to use models like `gpt-5.1-codex-max`, `gpt-5.1-codex-mini`, `gpt-5.2`, `gpt-5.2-codex` at no additional cost.

Reference: OpenCode's `codex.ts` plugin implements this flow.

## Key OAuth Details (from OpenCode)

- **OAuth Issuer:** `https://auth.openai.com`
- **Client ID:** `app_EMoamEEZ73f0CkXaXp7hrann`
- **API Endpoint:** `https://chatgpt.com/backend-api/codex/responses` (requests are rewritten to this URL)
- **Auth flow:** OAuth 2.0 Authorization Code + PKCE (S256)
- **Scopes:** `openid profile email offline_access`
- **Token endpoint:** `https://auth.openai.com/oauth/token`
- **Extra params:** `codex_cli_simplified_flow=true`, `originator=frontman`
- **Token response includes:** `access_token`, `refresh_token`, `id_token`, `expires_in`
- **Account ID:** Extracted from JWT claims (`chatgpt_account_id` or organization ID), sent as `ChatGPT-Account-Id` header
- **Allowed models:** `gpt-5.1-codex-max`, `gpt-5.1-codex-mini`, `gpt-5.2`, `gpt-5.2-codex`
- **URL rewriting:** All requests to `/v1/responses` or `/chat/completions` get rewritten to the Codex endpoint

### What can be reused from Anthropic integration

1. **`OAuthToken` schema** - Already provider-agnostic (has `provider` field). No changes needed.
2. **`Providers` context** - `upsert_oauth_token/5`, `get_oauth_token/2`, `get_valid_oauth_token/2`, `delete_oauth_token/2`, `has_oauth_token?/2` are all provider-agnostic. Minor changes needed for refresh routing.
3. **Client state types** - Generalize `anthropicOAuthStatus` to be per-provider or add `chatgptOAuthStatus` alongside it.
4. **Settings UI pattern** - Copy the Anthropic OAuth card in `Client__SettingsModal.res` and adapt for ChatGPT.
5. **Models controller pattern** - Add `@openai_codex_provider` alongside `@anthropic_provider`.

#### Server (Elixir)

1. **`ChatGPTOAuth` module** (mirrors `AnthropicOAuth`)
   - `generate_pkce/0` - Can extract to shared module (identical logic)
   - `build_authorize_url/2` - Different client_id, auth_url, scopes, redirect_uri
   - `exchange_code/2` - Different token_url, client_id, content-type (`application/x-www-form-urlencoded` vs JSON)
   - `refresh_token/1` - Different token_url, client_id, content-type
   - `calculate_expires_at/1` - Can reuse (identical logic)
   - **New:** `extract_account_id/1` - Parse JWT to extract `chatgpt_account_id`

2. **`ChatGPTOAuthController`** (mirrors `AnthropicOAuthController`)
   - Same 4 endpoints: authorize-url, exchange, disconnect, status
   - Store `account_id` in new column or metadata on `OAuthToken`

3. **Router** - Add `/api/oauth/chatgpt/*` routes

4. **`Providers` context changes:**
   - Update `maybe_resolve_oauth_token/2` to also handle `"openai"` provider
   - Return OpenAI-specific transformation opts (URL rewriting, Bearer auth, `ChatGPT-Account-Id` header)
   - Update `refresh_oauth_token/2` to dispatch to correct OAuth module based on provider

5. **`ModelsController`** - Add `@openai_codex_provider` with the allowed Codex models

6. **`LLMClient`** changes:
   - Handle URL rewriting to `chatgpt.com/backend-api/codex/responses` for OpenAI OAuth
   - Set `ChatGPT-Account-Id` header
   - Use Bearer token auth (like Anthropic OAuth)
   - No identity override or mcp_ prefix needed

7. **DB migration** - Add `metadata` column to `oauth_tokens` (for `account_id`)

#### Client (ReScript)

8. **State types** - Add `chatgptOAuthStatus` to state (parallel to `anthropicOAuthStatus`)
9. **State reducer** - Add actions/effects mirroring Anthropic ones but hitting `/api/oauth/chatgpt/*`
10. **Settings Modal** - Add ChatGPT Pro/Plus card in providers tab
11. **Models config** - Handle new OpenAI Codex provider in model selection

### Code reuse opportunities

- **Extract shared PKCE module** on server (both providers use identical PKCE logic)
- **Consider generic OAuth controller** - both controllers have identical structure, differing only in the OAuth module and provider string
- **Client-side:** Consider a generic `OAuthProviderCard` component for the settings modal
- **`hasAnyProviderConfigured` selector** - Add `chatgptOAuthStatus` check

## Auth Flow Difference

Unlike Anthropic (which uses manual code copy/paste because their callback goes to `console.anthropic.com`), OpenAI's flow can use a **local callback server** (port 1455, `localhost:1455/auth/callback`). This is a better UX (fully automatic). However, since Frontman is a web app (not CLI), we might need to use a server-side callback or the same manual code copy approach.

**Decision needed:** Use server-side OAuth callback (best UX) or manual code copy (consistent with Anthropic)?

### PR DESCRIPTION
**Pull request by itayadler**
## Summary

Adds support for connecting ChatGPT Pro/Plus subscriptions to use OpenAI Codex models (gpt-5.1-codex-max, gpt-5.1-codex-mini, gpt-5.2, gpt-5.2-codex) through Frontman. Uses the same OAuth device auth flow as the official Codex CLI.

- **Device Auth Flow**: User gets a code, enters it at `auth.openai.com/codex/device`, server polls for token completion
- **Codex API Routing**: Routes to `chatgpt.com/backend-api/codex/responses` (Responses API only, with `store:false`, `instructions` field, `ChatGPT-Account-Id` header)
- **Multi-turn tool calls**: Encodes full conversation history with `function_call`/`function_call_output` items inline for stateless Codex sessions (no `previous_response_id`)
- **Telemetry**: Fixes `String.Chars` crash when model is an `LLMDB.Model` struct

### ReqLLM Fork Changes
Three commits pushed to `frontman-ai/req_llm` `main`:
1. Extract system messages into `instructions` field for Responses API
2. Support `store` option in provider_options
3. **Encode full conversation history with function_call items when no previous_response_id** — critical for Codex multi-turn tool call support

Closes #269

---

<a href="https://app.devin.ai/review/frontman-ai/frontman/pull/289" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open with Devin
  </picture>
</a>

### CODE REVIEW

File: libs/client/src/state/Client__State__StateReducer.res [hunk 1]
```diff
@@ -728,6 +763,172 @@ let handleEffect = (effect, state: state, dispatch) => {
...
+          | _ =>
+            await Promise.make((resolve, _) => {
+              let _ = Js.Global.setTimeout(() => resolve(.), intervalMs)
+            })
+            await pollLoop(attempt + 1)
+          }
+        }
+      }
+      await pollLoop(0)
+    }
```
[the remark this answers is not in the document]
Comment (itayadler) [reply libs/client/src/state/Client__State__StateReducer.res:918]: Addressed in cfe7891 + 7bccef2. Rather than a mutable cancellation ref, we tag actions with `deviceAuthId` and the reducer only applies transitions when the ID matches the current `ChatGPTShowingCode` state. 8 tests in `Client__ChatGPTOAuth.test.res` cover all edge cases (matching/mismatched IDs, None for non-poll errors, full retry flow). Purely functional — no global mutable state.

File: apps/frontman_server/lib/frontman_server/agents/llm_client.ex [hunk 2]
```diff
@@ -190,8 +190,18 @@ defimpl Swarm.LLM, for: FrontmanServer.Agents.LLMClient do
...
-    Chunk.done(reason)
+  # Carry through response_id for Responses API previous_response_id chaining
+  defp to_swarm_chunk(
+         %{type: :meta, metadata: %{finish_reason: reason} = meta},
+         _requires_mcp_prefix?
+       ) do
+    chunk_meta =
+      case meta do
+        %{response_id: id} when is_binary(id) -> %{response_id: id}
+        _ -> %{}
```
Comment (BlueHotDog) [inline apps/frontman_server/lib/frontman_server/agents/llm_client.ex:201]: what are the other options here? what happens if we return this map?

File: apps/frontman_server/lib/frontman_server/agents/schema_transformer.ex [hunk 3]
```diff
@@ -56,11 +56,15 @@ defmodule FrontmanServer.Agents.SchemaTransformer do
   Returns `:openai_strict` for OpenAI and Azure models (via OpenRouter or direct),
   `:flexible` for all others.
   """
-  @spec provider_for_model(String.t()) :: provider()
+  @spec provider_for_model(String.t() | %{provider: atom()}) :: provider()
   def provider_for_model(model) when is_binary(model) do
```
Comment (BlueHotDog) [inline apps/frontman_server/lib/frontman_server/agents/schema_transformer.ex:60]: why do we support both atom and string? lets move to a single interface

File: apps/frontman_server/lib/frontman_server/observability/swarm_otel_handler.ex [hunk 4]
```diff
@@ -231,17 +231,18 @@ defmodule FrontmanServer.Observability.SwarmOtelHandler do
 
   @doc false
   def handle_llm_start(_event, _measurements, metadata, _config) do
-    %{loop_id: loop_id, step: step, model: model} = metadata
+    %{loop_id: loop_id, step: step, model: raw_model} = metadata
```
Comment (BlueHotDog) [inline apps/frontman_server/lib/frontman_server/observability/swarm_otel_handler.ex:234]: what is raw_model vs model?

File: apps/frontman_server/lib/frontman_server/providers/oauth_helpers.ex [hunk 5]
```diff
@@ -0,0 +1,41 @@
+defmodule FrontmanServer.Providers.OAuthHelpers do
```
Comment (BlueHotDog) [inline apps/frontman_server/lib/frontman_server/providers/oauth_helpers.ex:1]: i dont like helpers.. lets find a good place for these functions
